Getting KeyError: 'ERROR' when running hole-in-plate-quarter - fixes #56

jchkoch opened the pull request:
This PR solves the issue #56 by adding a new result data field. ccx solves for the error from each node and captures this in the results file (i.e. .frd file) which pycalculix `solve` function in `problem.py` would not understand. Therefore, I added a new mode and mode name with associated functions to make the results from the new mode be accessible.

spacether commented on pycalculix/base_classes.py:
```diff
@@ -25,6 +25,7 @@
 RESFIELDS['stress'] = 'Sx,Sy,Sz,Sxy,Syz,Szx,Seqv,S1,S2,S3'.split(',')
 RESFIELDS['strain'] = 'ex,ey,ez,exy,eyz,ezx,eeqv,e1,e2,e3'.split(',')
 RESFIELDS['force'] = 'fx,fy,fz'.split(',')
+RESFIELDS['error'] = 'STR(%)'.split(',')
```
Is this strain percent?
Could we give this a more descriptive name?

jchkoch commented:
I can definitely add a python test loading in a results file. I will have to find some time to do this. To give some background, I simply was trying to get started using the pycalculix and was working on getting the `hole-in-plate-full.py` example to run fully and I needed to add this. From what I can tell this was necessary since I installed ccx-2.16 (under opensuse leap 15.1).
I think that the error is the strain percent but I will have another look at ccx documentation.

jchkoch commented:
For your second question/ask regarding what "STR(%)" refers to, according to the documentation of ccx-2.16, the error that is calculated is:

>  ERR [ERROR (real), ERRORI (imaginary)]: error estimator for struc-
tural calculations (cf. Section 6.12). Notice that ERR and ZZS are mutu-
ally exclusive.

And further information from Section 6.12 of ccx-2.16 docs:

> The Zienkiewicz-Zhu error estimator [93], [94] tries to estimate the error made by the finite element discretization. To do so, it calculates for each node an improved stress and defines the error as the difference between this stress and the one calculated by the standard finite element procedure.

I have updated the documentation string at the top of `base_classes.py` to explain what the field is with a more descriptive explanation. Additionally, I have changed the code to read as you suggested `RESFIELDS['error'] = ['STR(%)']`.

Then to address your first ask about adding a results file and python test for loading it, I can definitely do that although I am not sure it is strictly necessary. I ran the existing unittests which ran calling version 2.16 of ccx which creates the "error" result block in the results file and they all run successfully (I did have to specify the full path to the example python files in the function `example_tester`). Please advise @spacether if you would still like me to include in a commit the results frd file (it is quite a large file).

The OS that I am using is `openSUSE Leap 15.1` and I installed calculix from the openSUSE repository [`science`](https://en.opensuse.org/openSUSE:Science_Repositories) Let me know if you need anything further from me.

jchkoch commented:
just added a test to prove that loading the results file I get using ccx version 2.16 works on my opensuse leap 15.1 setup. The results file is `examples/hole-in-plate-full-from-ccx-ver2-16.frd`. Added the test in the file test_samples.py as it seemed logical to include it in that file (i.e. not create a new test_*.py file).

spacether commented on pycalculix/base_classes.py:
```diff
@@ -8,11 +8,13 @@
             - 'stress': stress
             - 'strain': strain
             - 'force': force
+            - 'error': FE discretization error in stress
```
Should we change this to 'error in strain`?

spacether commented on pycalculix/base_classes.py:
```diff
@@ -8,11 +8,13 @@
             - 'stress' = 'Sx','Sy','Sz','Sxy','Syz','Szx','Seqv','S1','S2','S3'
             - 'strain' = 'ex','ey','ez','exy','eyz','ezx','eeqv','e1','e2','e3'
             - 'force' = 'fx','fy','fz'
+            = 'error' = 'STR(%)'
```
Can we change this to e% to be consistent with our strain nomenclature?
Or strain% to be fully explicit?
Let's also keep this docstring consistent with our code below.

spacether commented on pycalculix/base_classes.py:
```diff
@@ -25,6 +27,7 @@
 RESFIELDS['stress'] = 'Sx,Sy,Sz,Sxy,Syz,Szx,Seqv,S1,S2,S3'.split(',')
 RESFIELDS['strain'] = 'ex,ey,ez,exy,eyz,ezx,eeqv,e1,e2,e3'.split(',')
 RESFIELDS['force'] = 'fx,fy,fz'.split(',')
+RESFIELDS['error'] = ['STR(%)']
```
Can we change this to e% to be consistent with our strain nomenclature?
Or strain% to be fully explicit?

spacether commented on pycalculix/results_file.py:
```diff
@@ -1077,6 +1078,16 @@ def _save_node_force(self, line, rfstr, time, mode='force'):
 
+    def _save_node_error(self, line, rfstr, time, mode='error'):
+        """Saves node error"""
+        # [key, node, error]
```
Can you please add a comment here describing that the error value
crom the ccx docs? I want the devs to know that the error is a percent difference in strain.

spacether commented:
Thank you for the PR! Looks good.
